File: reverseLinkedList.js
/* Hooray! It's opposite day. Linked lists go the opposite way today.
Write a function for reversing a linked list. ↴ Do it in-place. ↴

Your function will have one input: the head of the list.

Your function should return the new head of the list.

Here's a sample linked list node class:

  function LinkedListNode(value) {
    this.value = value;
    this.next = null;
  }
*/

const reverseList = (node) => {
  let current = node;
  let prev = null;
  let nextNode = null;
  while (current) {
    // capture next node
    nextNode = current.next;
    // assign previous
    current.next = prev;
    // move to next node
    prev = current;
    current = nextNode;
  }
  return prev;
};

function LinkedListNode(value) {
  this.value = value;
  this.next = null;
}

let linkedList = new LinkedListNode(0);
linkedList.next = new LinkedListNode(1);


console.log(linkedList);

reverseLinkedList(linkedList);

File: reverseLinkedList.py
linkedlist = {
  'val': 0,
  'next': {
    'val': 1,
    'next': {
      'val': 2,
      'next': None
    }
  }
}

# Input: linked list head
# Output: reversed list head

# Skeleton
# - while curr exists or does not equal None
# - if there's no existing result
# -- store the current node's next value
# -- make the current node prev and result_head
# -- make current node's value None 
# -- make next the new current node 
# - else
# -- store current node's next value
# -- make current node's next value equal prev
# -- make prev equal current node
# -- make curr equal next
# - return reversed list head


class Solution:
  def reverseList(self, l):
    print('HELLO')
    

test = Solution()
print(test.reverseList(linkedlist))